underfs/ssh/src/main/java/alluxio/underfs/ssh/SftpFactory.java

```diff
@@ -0,0 +1,351 @@
+  private static ScheduledExecutorService sExecutor = Executors.newSingleThreadScheduledExecutor(
+      ThreadFactoryUtils.build("SftpFactory-close-idle-sftp-%d", true));
+
+  /**
+   * Create a sftp client.
+   * The returned instance is a proxy of SFTPClient in order to support auto connection
+   * and release feature. The ssh and sftp will be connected if the member method called first time,
+   * and the sftp and ssh will be closed if the instance is not used for a while.
+   *
+   * @param uri the {@link AlluxioURI} for this UFS
+   * @param ufsConf UFS configuration
+   * @return the generated instance
+   */
+  public static SFTPClient createSftpClient(AlluxioURI uri, UnderFileSystemConfiguration ufsConf) {
+    SftpContainer container = new SftpContainer(uri, ufsConf);
+    sLiveContainers.add(container);
+    return container.getSftpClientProxy();
+  }
+
+  /**
+   * Transform a remote file to an input stream, which will close the RemoteFile instance
+   * in close method. And the remote file should not be accessed after transformation.
+   *
+   * @param remoteFile the remote file
+   * @param offset the offset
+   * @return the input stream
+   * @throws IOException
+   */
+  public static InputStream transToInputStream(RemoteFile remoteFile, long offset)
+      throws IOException {
+    Enhancer enhancer = new Enhancer();
+    enhancer.setCallback(new CloseInterceptor(remoteFile));
+    enhancer.setSuperclass(RemoteFile.ReadAheadRemoteFileInputStream.class);
+
+    // The returned instance is a proxy of remoteFile.new ReadAheadRemoteFileInputStream(int, long),
+    // and the proxy behavior is to close the remoteFile in close method
+    return (InputStream) enhancer.create(
+        new Class[] {RemoteFile.class, int.class, long.class},
+        new Object[] {remoteFile, STREAM_CACHE_RW_NUM, offset});
+  }
+
+  /**
+   * Transform a remote file to an output stream, which will close the RemoteFile instance
+   * in close method. And the remote file should not be accessed after transformation.
+   *
+   * @param remoteFile the remote file
+   * @return the output stream
+   * @throws IOException
+   */
+  public static OutputStream transToOutputStream(RemoteFile remoteFile)
+      throws IOException {
+    Enhancer enhancer = new Enhancer();
+    enhancer.setCallback(new CloseInterceptor(remoteFile));
+    enhancer.setSuperclass(RemoteFile.RemoteFileOutputStream.class);
+
+    // The returned instance is a proxy of remoteFile.new RemoteFileOutputStream(long, int),
+    // and the proxy behavior is to close the remoteFile in close method
+    return (OutputStream) enhancer.create(
+        new Class[] {RemoteFile.class, long.class, int.class},
+        new Object[] {remoteFile, 0, STREAM_CACHE_RW_NUM});
+  }
+
+  static {
+    long delay =  MAX_IDLE_INTERVAL / 2;
+    sExecutor.scheduleAtFixedRate(() -> scanIdle(), delay, delay, TimeUnit.MILLISECONDS);
+  }
+
+  protected static void scanIdle() {
```

Comment:
is this actually close idle connections. if it is please rename this method

Reply:
closeSftp is renamed to closeIfIdle, so I think this method name should not be changed